crates/ruma-events/src/room/join_rules.rs

```diff
@@ -178,6 +178,59 @@ impl Restricted {
     }
 }
 
+impl<'de> Deserialize<'de> for Restricted {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        struct S {
+            allow: IgnoreInvalidItems<AllowRule>,
+        }
+
+        let S { allow: IgnoreInvalidItems(allow) } = Deserialize::deserialize(deserializer)?;
+
+        Ok(Self { allow })
+    }
+}
+
+/// Deserialize a `Vec<T>`, dropping all `T` that don't match the expected schema
+struct IgnoreInvalidItems<T>(Vec<T>);
+
+impl<'de, T> Deserialize<'de> for IgnoreInvalidItems<T>
```

Comment:
I imagine this stuff should probably be defined somewhere other than right here, but I'm not sure where that would be.